keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2300 @@
+annotations will block eviction requests.
+
+`kubectl cordon` and `kubectl uncordon` commands will be enhanced with a warning that will warn
+the user if a node is made un/schedulable, and it collides with an existing NodeMaintenance object.
+As a consequence the node maintenance controller will reconcile the node back to the old value.
+Because of this we can make these commands noop when the node is under the NodeMaintenance.
+
+### NodeMaintenance API
+
+NodeMaintenance objects serve as an intent to remove or migrate pods from a set of nodes. We will
+include Cordon and Drain toggles to support the following states/stages of the maintenance:
+1. Planning: this is to let the users know that maintenance will be performed on a particular set
+   of nodes in the future. Configured with `.spec.stage=Idle`.
+2. Cordon: stop accepting (scheduling) new pods. Configured with `.spec.stage=Cordon`.
+3. Drain: gives an intent to drain all selected nodes by creating `Evacuation` objects for the
+   node's pods. Configured with `.spec.stage=Drain`.
+4. Drain Complete: all targeted pods have been drained from all the selected nodes. The nodes can
+   be upgraded, restarted, or shut down. The configuration is still kept at `.spec.stage=Drain` and
+   `Drained` condition is set to `"True"` on the node maintenance object.
+5. Maintenance Complete: make the nodes schedulable again once the node maintenance is done.
+   Configured with `.spec.stage=Complete`.
+
+```golang
+
+// +enum
+type NodeMaintenanceStage string
+
+const (
+// Idle does not interact with the cluster.
+Idle NodeMaintenanceStage = "Idle"
+// Cordon cordons all selected nodes by making them unschedulable.
+Cordon NodeMaintenanceStage = "Cordon"
+// Drain:
+// 1. Cordons all selected nodes by making them unschedulable.
+// 2. Gives an intent to drain all selected nodes by creating Evacuation objects for the
+//    node's pods.
+Drain NodeMaintenanceStage = "Drain"
+// Complete:
+// 1. Removes all Evacuation objects requested by this NodeMaintenance.
+// 2. Uncordons all selected nodes by making them schedulable again, unless there is not another
+//    maintenance in progress.
+Complete NodeMaintenanceStage = "Complete"
+)
+
+type NodeMaintenance struct {
+    ...
+    Spec NodeMaintenanceSpec
+    Status NodeMaintenanceStatus
+}
+
+type NodeMaintenanceSpec struct {
+    // NodeSelector selects nodes for this node maintenance.
+    // +required
+    NodeSelector *v1.NodeSelector
+
+    // The order of the stages is Idle -> Cordon -> Drain -> Complete.
+    //
+    // - The Cordon or Drain stage can be skipped by setting the stage to Complete.
+    // - The NodeMaintenance object is moved to the Complete stage on deletion unless the Idle stage has been set.
+    //
+    // The default value is Idle.
+    Stage NodeMaintenanceStage
+
+    // DrainPlan is executed from the first entry to the last entry during the Drain stage.
```

Comment:
> If this is a must, what if there are dependencies between Pods that don't follow this order? (e.g. a static Pod depending on a DaemonSet, or a DaemonSet depending on a Default Pod)

It is must. And I think basically all clusters follow this ordering today. Do you have an example where this is not the case?

> Similar question. Is this just a should or will the ordering based on priority per podType be enforced?

enforced

> This implies that the stages are executed based on this order and not the order of the entries. (But this would make sense if the order of the entries are enforced to have this ordering)

the order of the entries should conform to the type and priority ordering. A lower index entry with a `podSelector` and the same priority will be drained first.

Reply:
> It is must. And I think basically all clusters follow this ordering today. Do you have an example where this is not the case?

Sorry no concrete example. I was only wondering if it's safe to assume that the order in which Pods should be drained can simply be determined by the way that they are deployed (Deployments/operator/... vs DaemonSet). The way that Pods are deployed and drain order seem orthogonal to me.